ivadomed/loader/segmentation_pair.py

```diff
@@ -315,12 +315,12 @@ def convert_file_to_nifti(self, filename: str, extension: str, is_gt: bool = Fal
         """
         # For '.png', '.tif', '.tiff', '.jpg' and 'jpeg' extentions
         # Read image as 8 bit grayscale in numpy array (behavior TBD in ivadomed for RGB, RBGA or higher bit depth)
-        if "tif" in extension:
-            img = np.expand_dims(imageio.v2.imread(filename, format='tiff-pil'), axis=-1).astype(np.uint8)
-            if len(img.shape) > 3:
-                img = np.expand_dims(imageio.v2.imread(filename, format='tiff-pil', as_gray=True), axis=-1).astype(np.uint8)
-        else:
-            img = np.expand_dims(imageio.v2.imread(filename, as_gray=True), axis=-1).astype(np.uint8)
+        _img = imageio.v3.imread(filename)
```

Comment:
If all you want is 8-bit grayscale for PNG, JPEG, and TIFF, why not use

```python
_img = imageio.v3.imread(filename, plugin="pillow", mode="L")
```

Pillow's TIFF support is much less extensive than tifffile's TIFF support, but assuming that you have simple TIFFs pillow should be able to handle this.

(the diff here uses `TIFF-PIL`, which suggests that this is indeed sufficient.)

Reply:
Essentially what we are trying to do is accomodate both 8-bits and 16-bits for PNG/TIFF files. Using `mode="F"` works for 8bit and 16bit PNG but pillow is not able to read our TIFFs. Hence, we need to switch to tifffile for TIFF's to be loaded properly

```python
 if 'tif' in extension:
   _img = imageio.v3.imread(filename, plugin='tifffile')
else:
   _img = imageio.v3.imread(filename, plugin="pillow", mode="F")
``` 
Now this here _almost_ works. Only pitfall is that we would like to read rgb/rgba images as grayscale images. The simplest fix I can think of at this point is converting the rgb tiff image to grayscale when we see that a tiff image has more than 1 channel.

Reply:
> I get this error trying to read a simple 8bit TIFF

Interesting. Pillow doesn't get the filepointer for the file that you are trying to read. It could be that this isn't TIFF specific, but rather an issue with how ImageIO and pillow interact. What type of ImageResource are you loading? (file, stream, `bytes`, ...)

If you are able to share an offending TIFF I should be able to run it locally and check if it is perhaps a bug in ImageIO v3. It reminds me of a report for a different format I got a while ago; unfortunately I never managed to reproduce the issue, so I could never figure out what the problem was.

> Another problem is that we often have users with 16 bit images [...] Only pitfall is that we would like to read rgb/rgba images as grayscale images.

I see. In this case I would approach the problem differently. Rather than solving the colorspace conversion indirectly during load by constraining the resulting dtype I would separate the loading (decoding and conversion to `np.array`) and processing (colorspace conversion, and/or buffer unification) steps:

```python
import imageio.v3 as iio

props = iio.improps(filename). # new in v3 - standardized metadata
_img = iio.imread(filename)

# TIFF is a "wild" format. A few assumptions greatly simplify the code below:
# 1. the image is interleaved/channel-last (not planar)
# 2. the colorspace is one of: binary, gray, RGB, RGBA (not aliasing ones like YUV or CMYK)
# 3. the image is flat (not a volume or time-series)
# Note: All of these are trivially true for JPEG and PNG due to limitations of these formats.

# make grayscale (treats binary as 1-bit grayscale)
colorspace_idx = 2 + int(props.is_batch)
if _img.ndims <= colorspace_idx:  # binary or gray
    pass  # nothing to do
elif _img.shape[colorspace_idx] == 3:  # RGB
    _img = np.sum(_img * (.299, .587, .114), axis=-1)
else:  # RGBA
    # discards alpha
    _img = np.sum(_img * (.299, .587, .114, 0), axis=-1)

```

This version lets ImageIO decide how to do the loading (via pillow, tifffile, magic, etc.).

Reply:
> > I get this error trying to read a simple 8bit TIFF
> 
> Interesting. Pillow doesn't get the filepointer for the file that you are trying to read. It could be that this isn't TIFF specific, but rather an issue with how ImageIO and pillow interact. What type of ImageResource are you loading? (file, stream, `bytes`, ...)

It is a file. This is the image I was talking about: https://github.com/ivadomed/data-testing/blob/0f4d5b644397bf4ab2f4b02f1e30065d4aaba61e/data_test_png_tif/sub-tif/micr/sub-tif_sample-8gray_SEM.tif

> In this case I would approach the problem differently. Rather than solving the colorspace conversion indirectly during load by constraining the resulting dtype I would separate the loading (decoding and conversion to `np.array`) and processing (colorspace conversion, and/or buffer unification) steps:
> 
> ```python
> import imageio.v3 as iio
> 
> props = iio.improps(filename). # new in v3 - standardized metadata
> _img = iio.imread(filename)
> 
> # TIFF is a "wild" format. A few assumptions greatly simplify the code below:
> # 1. the image is interleaved/channel-last (not planar)
> # 2. the colorspace is one of: binary, gray, RGB, RGBA (not aliasing ones like YUV or CMYK)
> # 3. the image is flat (not a volume or time-series)
> # Note: All of these are trivially true for JPEG and PNG due to limitations of these formats.
> 
> # make grayscale (treats binary as 1-bit grayscale)
> colorspace_idx = 2 + int(props.is_batch)
> if _img.ndims <= colorspace_idx:  # binary or gray
>     pass  # nothing to do
> elif _img.shape[colorspace_idx] == 3:  # RGB
>     _img = np.sum(_img * (.299, .587, .114), axis=-1)
> else:  # RGBA
>     # discards alpha
>     _img = np.sum(_img * (.299, .587, .114, 0), axis=-1)
> ```
> 
> This way ImageIO can decide how to do the loading (via pillow, tifffile, magic, etc.). You don't have to create different code paths based on the image format nor do you have to worry about the plugin that is used to do the decoding.

I actually love this solution!